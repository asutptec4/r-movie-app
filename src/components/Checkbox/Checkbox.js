import { bool, func, string } from 'prop-types';
import React from 'react';

import { useComponentDidUpdate, useToggle } from '../../utils/custom-hooks';
import { defaultHandler } from '../../utils/util-func';

import './Checkbox.scss';

const Checkbox = ({ label, initialState, handleOnChange, className }) => {
  const [isChecked, toggle] = useToggle(initialState);

  useComponentDidUpdate(() => {
    handleOnChange(isChecked);
  }, [isChecked]);

  return (
    <label className={`checkbox ${className}`.trim()}>
      {label}
      <input type="checkbox" checked={isChecked} onChange={toggle} />
      <span className="checkmark"></span>
    </label>
  );
};

Checkbox.defaultProps = {
  className: '',
  initialState: false,
  handleOnChange: defaultHandler,
};

Checkbox.propTypes = {
  label: string,
  className: string,
  initialState: bool,
  handleOnChange: func,
};

export default Checkbox;
