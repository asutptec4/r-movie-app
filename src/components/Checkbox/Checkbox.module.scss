@import '../../scss/index.scss';

.checkbox {
    display: block;
    position: relative;
    padding-left: 1.5em;
    cursor: pointer;
    user-select: none;

    &:hover input ~ .checkmark {
      background-color: $item-color;
    }

    input {
      position: absolute;
      opacity: 0;
      cursor: pointer;
      height: 0;
      width: 0;

      &:checked ~ .checkmark {
        background-color: $item-color;

        &:after {
          display: block;
        }
      }
    }

    .checkmark {
      position: absolute;
      top: 0;
      left: 0;
      height: 1em;
      width: 1em;
      background-color: $label-color;

      &:after {
        content: '';
        position: absolute;
        display: none;
        left: 0.3em;
        top: 0.1em;
        width: 0.25em;
        height: 0.5em;
        border: solid $label-color;
        border-width: 0 3px 3px 0;
        transform: rotate(45deg);
      }
    }
  }