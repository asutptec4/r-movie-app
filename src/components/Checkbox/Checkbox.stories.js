import { withKnobs, text } from '@storybook/addon-knobs';
import React from 'react';

import Checkbox from './Checkbox';

export default {
  title: 'MovieApp/Checkbox',
  component: Checkbox,
  parameters: { actions: { argTypesRegex: '^handle.*' } },
  decorators: [withKnobs],
};

const Template = (args) => <Checkbox {...args} />;

export const Checked = Template.bind({});
Checked.args = {
  label: 'Checkbox Label',
  initialState: true,
};

export const Unchecked = Template.bind({});
Unchecked.args = {
  label: 'Checkbox Label',
  initialState: false,
};

export const CustomClassName = Template.bind({});
CustomClassName.args = {
  initialState: true,
  label: 'Checkbox Label',
  className: 'custom',
};

export const CustomLabel = () => <Checkbox label={text('Checkbox Label', 'label')} initialState="true" />;
