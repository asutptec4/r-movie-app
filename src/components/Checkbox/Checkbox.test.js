import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import Checkbox from './Checkbox';

describe('<Checkbox />', () => {
  const mockHandler = jest.fn();
  const defaultProps = { label: 'Checkbox', initialState: true, className: 'customClass', handleOnChange: mockHandler };

  test('should render props', () => {
    const { container } = render(<Checkbox {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('should change state by click and invoke handler', () => {
    const { getByLabelText } = render(<Checkbox {...defaultProps} />);
    const input = getByLabelText(defaultProps.label);
    expect(input.checked).toBeTruthy();
    fireEvent.click(input);
    expect(input.checked).toBeFalsy();
    expect(mockHandler).toHaveBeenCalledWith(false);
  });
});
