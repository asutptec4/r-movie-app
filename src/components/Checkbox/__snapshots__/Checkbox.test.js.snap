// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render props 1`] = `
<div>
  <label
    class="checkbox customClass"
  >
    Checkbox
    <input
      checked=""
      type="checkbox"
    />
    <span
      class="checkmark"
    />
  </label>
</div>
`;
